File: Csv-Files/MapGenerator.cpp
#include <iostream>
#include <fstream>
#include <time.h>
#include <queue>

#define GRIDSIZE 500
#define PERCENTBUILDINGS 20
#define RANGEOFCRIMEVALUES 9 //works up to 9 only due to "char" nature of this program which originally did not handle digits
#define OUTPUTFILENAME "randomizedCityMap.csv"



bool checkIfNotEnclosed(char storeArray[GRIDSIZE][GRIDSIZE], int, int);
void floodfill(char storeArray[GRIDSIZE][GRIDSIZE], int, int);

int main()
{
    srand(time(NULL));

    char storeArray[GRIDSIZE][GRIDSIZE * 2];


    //randomize data
    for (int i = 0; i < GRIDSIZE; i++)
    {
        for (int j = 0; j < GRIDSIZE; j++)
        {
            //1 through 100
            int randomNumber = rand() % 100 + 1;

            if (randomNumber <= PERCENTBUILDINGS)
            {
                storeArray[i][j] = '0';
            }
            else
            {
                int randomInt = rand() % RANGEOFCRIMEVALUES + 1;
                storeArray[i][j] = randomInt + '0';
            }
        }
    }
    //make borders clear
    for (int i = 0; i < GRIDSIZE; i++)
    {
        int randomInt = rand() % RANGEOFCRIMEVALUES + 1;
        storeArray[0][i] = randomInt + '0';
        randomInt = rand() % RANGEOFCRIMEVALUES + 1;
        storeArray[GRIDSIZE - 1][i] = randomInt + '0';
        randomInt = rand() % RANGEOFCRIMEVALUES + 1;
        storeArray[i][0] = randomInt + '0';
        randomInt = rand() % RANGEOFCRIMEVALUES + 1;
        storeArray[i][GRIDSIZE - 1] = randomInt + '0';
    }

    char temp[GRIDSIZE][GRIDSIZE];

    for (int i = 0; i < GRIDSIZE; i++)
    {
        for (int j = 0; j < GRIDSIZE; j++)
        {
            temp[i][j] = storeArray[i][j];
        }
    }
    //escape out enclosed streets
    for (int i = 1; i < GRIDSIZE - 1; i++)
    {
        for (int j = 1; j < GRIDSIZE - 1; j++)
        {
            if (temp[i][j] == '0' || temp[i][j] == 'A')
                continue;

            //reset temp array
            for (int i = 0; i < GRIDSIZE; i++)
            {
                for (int j = 0; j < GRIDSIZE; j++)
                {
                    temp[i][j] = storeArray[i][j];
                }
            }

            int shiftCount = 1;
            while (!checkIfNotEnclosed(temp, i, j))
            {
                std::cout << "Found index " << i << " " << j << " that is completely enclosed, blindly opening up one block to the left and checking again.\n";

                int randomInt = rand() % RANGEOFCRIMEVALUES + 1;
                storeArray[i - shiftCount][j] = randomInt + '0';
                shiftCount++;
                //reset temp array
                for (int i = 0; i < GRIDSIZE; i++)
                {
                    for (int j = 0; j < GRIDSIZE; j++)
                    {
                        temp[i][j] = storeArray[i][j];
                    }
                }
            }
        }
    }


    //put in commas
    for (int i = 0; i < GRIDSIZE; i++)
    {
        for (int j = GRIDSIZE - 1; j != -1; j--)
        {
            storeArray[i][2 * j] = storeArray[i][j];
            storeArray[i][2 * j + 1] = ',';
        }
    }
    

    //print out
    std::ofstream outFile;
    outFile.open(OUTPUTFILENAME);
    for (int i = 0; i < GRIDSIZE; i++)
    {
        for (int j = 0; j < GRIDSIZE * 2; j++)
        {
            outFile << storeArray[i][j];
        }
        outFile << "\n";
    }
    outFile.close();
    

    return 0;
}

bool checkIfNotEnclosed(char storeArray[GRIDSIZE][GRIDSIZE], int i, int j)
{
    floodfill(storeArray, i, j);

    if (storeArray[0][0] == 'A')
        return true;
    else
        return false;
}

void floodfill(char storeArray[GRIDSIZE][GRIDSIZE], int i, int j)
{
    if (storeArray[i][j] == '0' || storeArray[i][j] == 'A')
    {
        return;
    }
    else
    {
        storeArray[i][j] = 'A';
    }
    std::queue<int> xStack;
    std::queue<int> yStack;

    if (i - 1 >= 0)
    {
        xStack.push(i - 1);
        yStack.push(j);
    }
    if (i + 1 < GRIDSIZE)
    {
        xStack.push(i + 1);
        yStack.push(j);
    }
    if (j - 1 >= 0)
    {
        xStack.push(i);
        yStack.push(j-1);
    }
    if (j + 1 < GRIDSIZE)
    {
        xStack.push(i);
        yStack.push(j + 1);
    }

    while (!xStack.empty())
    {
        if (storeArray[xStack.front()][yStack.front()] == 'A' || storeArray[xStack.front()][yStack.front()] == '0')
        {
            xStack.pop();
            yStack.pop();
        }
        else
        {
            storeArray[xStack.front()][yStack.front()] = 'A';

            if (xStack.front() - 1 >= 0)
            {
                xStack.push(xStack.front() - 1);
                yStack.push(yStack.front());
            }
            if (xStack.front() + 1 < GRIDSIZE)
            {
                xStack.push(xStack.front() + 1);
                yStack.push(yStack.front());
            }
            if (yStack.front() - 1 >= 0)
            {
                xStack.push(xStack.front());
                yStack.push(yStack.front() - 1);
            }
            if (yStack.front() + 1 < GRIDSIZE)
            {
                xStack.push(xStack.front());
                yStack.push(yStack.front() + 1);
            }

            xStack.pop();
            yStack.pop();
        }
    }
}

File: Csv-Files/MapGenerator.py
import csv
import random
import time


def floodfill(storeArray, i, j):
    GRIDSIZE = 500
    if storeArray[i][j] == '0' or storeArray[i][j] == 'A':
        return
    else:
        storeArray[i][j] = 'A'

    xStack = []
    yStack = []

    if i - 1 >= 0:
        xStack.append(i - 1)
        yStack.append(j)
    if i + 1 < GRIDSIZE:
        xStack.append(i + 1)
        yStack.append(j)
    if j - 1 >= 0:
        xStack.append(i)
        yStack.append(j - 1)
    if j + 1 < GRIDSIZE:
        xStack.append(i)
        yStack.append(j + 1)

    while len(xStack) > 0:
        #front of queue is at 0 index
        if storeArray[xStack[0]][yStack[0]] == 'A' or storeArray[xStack[0]][yStack[0]] == '0':
            xStack.pop()
            yStack.pop()
        else:
            storeArray[xStack[0]][yStack[0]] = 'A'

            if xStack[0] - 1 >= 0:
                xStack.append(xStack[0] - 1)
                yStack.append(yStack[0])
            if xStack[0] + 1 < GRIDSIZE:
                xStack.append(xStack[0] + 1)
                yStack.append(yStack[0])
            if yStack[0] - 1 >= 0:
                xStack.append(xStack[0])
                yStack.append(yStack[0] - 1)
            if yStack[0] + 1 < GRIDSIZE:
                xStack.append(xStack[0])
                yStack.append(yStack[0] + 1)

            xStack.pop()
            yStack.pop()
            # print("lol")
    # print("Done")


def checkIfNotEnclosed(storeArray, i, j):
    floodfill(storeArray, i, j)

    if storeArray[0][0] == 'A':
        return True
    else:
        return False


if __name__ == '__main__':
    GRIDSIZE = 500
    PERCENTBUILDINGS = 20
    RANGEOFCRIMEVALUES = 9
    random.seed(time.time())  # equivalent to srand(time(NULL)) in c++
    # initalize 2D array
    storeArray = [['' for x in range(GRIDSIZE)] for y in range(GRIDSIZE * 2)]
    """ #testing
    for i in range(GRIDSIZE):
        for j in range(GRIDSIZE):
            print('{:500}'.format(storeArray[i][j]))
        print("\n")
    """

    for i in range(0, GRIDSIZE):
        for j in range(0, GRIDSIZE):
            randomNumber = random.randint(1, 101)
            # print(randomNumber,"\n")
            if randomNumber <= PERCENTBUILDINGS:
                storeArray[i][j] = '0'
            else:
                randomInt = random.randint(1, 11)
                storeArray[i][j] = str(randomInt) + '0'

    # Declaring borders
    for i in range(0, GRIDSIZE):
        randomInt = random.randint(1, 11)
        storeArray[0][i] = str(randomInt) + '0'
        randomInt = random.randint(1, 11)
        storeArray[GRIDSIZE - 1][i] = str(randomInt) + '0'
        randomInt = random.randint(1, 11)
        storeArray[i][GRIDSIZE - 1] = str(randomInt) + '0'

    # Initializing temp 2D array
    temp = [['' for x in range(GRIDSIZE)] for y in range(GRIDSIZE)]  # setting all to '0'

    for i in range(0, GRIDSIZE):
        for j in range(0, GRIDSIZE):
            temp[i][j] = storeArray[i][j]  # setting to storeArray values
    # BUG IS SOMEWHERE HERE


    for i in range(1, GRIDSIZE - 1):
        for j in range(1, GRIDSIZE - 1):
            if temp[i][j] == '0' or temp[i][j] == 'A':
                continue
            # reset temp array
            for i in range(0, GRIDSIZE):
                for j in range(0, GRIDSIZE):
                    temp[i][j] = storeArray[i][j]
            # print("i ", i, " j ", j, "\n")
            #print("before i ", i, " j ", j, "\n")
            shiftCount = 1

            # Here is the bug,
            while not checkIfNotEnclosed(temp, i, j):
                print("Found index ", i, " ", j,
                      "that is completely enclosed, blindly opening up one block to the left and checking again.\n ")
                randomInt = random.randint(1, 11)
                storeArray[i - shiftCount][j] = str(randomInt) + '0'
                shiftCount += 1

                for i in range(0, GRIDSIZE):
                    for j in range(0, GRIDSIZE):
                        temp[i][j] = storeArray[i][j]


    for i in range(0, GRIDSIZE):
        j = GRIDSIZE - 1
        while j != -1:
            storeArray[i][2 * j] = storeArray[i][j]
            storeArray[i][2 * j + 1] = ','
            j -= 1

    with open('newRandomMap.csv', mode='w') as newRandomMap:
        for i in range(0, GRIDSIZE):
            for j in range(0, GRIDSIZE * 2 - 1):
                newRandomMap = storeArray[i][j]
            newRandomMap = "\n"
